runtimes/rust/src/api.rs

```diff
@@ -244,27 +251,32 @@ impl ServiceApiScope {
     where
         Q: DeserializeOwned + 'static,
         I: Serialize + 'static,
-        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
-        R: IntoFuture<Item = I, Error = Error> + 'static,
+        F: Fn(ServiceApiState, Q) -> R + 'static + Clone + Send + Sync,
+        R: Future<Output = exonum_api::Result<I>>,
     {
         let blockchain = self.blockchain.clone();
         let descriptor = self.descriptor.clone();
         let artifact = self.artifact.clone();
-        self.inner
-            .endpoint_mut(name, move |query: Q| -> FutureResult<I> {
+        self.inner.endpoint_mut(name, move |query: Q| {
+            let handler = handler.clone();
+            let blockchain = blockchain.clone();
+            let descriptor = descriptor.clone();
+            let artifact = artifact.clone();
+
+            async move {
```

Comment:
I'd suggest to rework this to "`futures 0.1`-style" (i.e., minimized or absent `async {}`) in order to avoid cloning above. Can be done later.